File: packages/engine/src/ownership.ts
import { FILES } from "./render.js";

export const SERVICES = {
  launcher: "normal-launcher",
  appd: "normal-appd",
  notifyd: "normal-notifyd",
  attentiond: "normal-attentiond",
  webviewShim: "normal-webview-shim",
} as const;

export type ServiceName = (typeof SERVICES)[keyof typeof SERVICES];

export type FileOwnership = {
  readonly file: string;
  readonly sources: readonly string[];
  readonly readers: readonly ServiceName[];
};

export const FILE_OWNERSHIP: readonly FileOwnership[] = [
  {
    file: FILES.metadata,
    sources: ["/apiVersion", "/kind", "/metadata"],
    readers: [],
  },
  {
    file: FILES.launcher,
    sources: ["/spec/launcher"],
    readers: [SERVICES.launcher],
  },
  {
    file: FILES.apps,
    sources: ["/spec/apps"],
    readers: [SERVICES.appd, SERVICES.launcher],
  },
  {
    file: FILES.notifications,
    sources: ["/spec/notifications"],
    readers: [SERVICES.notifyd],
  },
  {
    file: FILES.attention,
    sources: ["/spec/attention"],
    readers: [SERVICES.attentiond],
  },
  {
    file: FILES.webviewShim,
    sources: ["/spec/attention", "/spec/apps"],
    readers: [SERVICES.webviewShim],
  },
];

const covers = (source: string, path: string): boolean =>
  path === source || path.startsWith(`${source}/`);

export const filesFor = (path: string): readonly string[] =>
  FILE_OWNERSHIP.filter((entry) => entry.sources.some((source) => covers(source, path))).map(
    (entry) => entry.file,
  );

export const isOwnedPath = (path: string): boolean => filesFor(path).length > 0;

export const readersOf = (file: string): readonly ServiceName[] =>
  FILE_OWNERSHIP.find((entry) => entry.file === file)?.readers ?? [];

export const servicesForFiles = (files: readonly string[]): readonly ServiceName[] =>
  [...new Set(files.flatMap(readersOf))].sort();

export const servicesFor = (paths: readonly string[]): readonly ServiceName[] =>
  servicesForFiles(paths.flatMap(filesFor));

File: pkg/engine/ownership.go
package engine

import (
	"sort"
	"strings"
)

const (
	ServiceLauncher    = "normal-launcher"
	ServiceAppd        = "normal-appd"
	ServiceNotifyd     = "normal-notifyd"
	ServiceAttentiond  = "normal-attentiond"
	ServiceWebviewShim = "normal-webview-shim"
)

type FileOwnership struct {
	File    string
	Sources []string
	Readers []string
}

func FileOwnerships() []FileOwnership {
	return []FileOwnership{
		{File: FileMetadata, Sources: []string{"/apiVersion", "/kind", "/metadata"}, Readers: nil},
		{File: FileLauncher, Sources: []string{"/spec/launcher"}, Readers: []string{ServiceLauncher}},
		{File: FileApps, Sources: []string{"/spec/apps"}, Readers: []string{ServiceAppd, ServiceLauncher}},
		{File: FileNotifications, Sources: []string{"/spec/notifications"}, Readers: []string{ServiceNotifyd}},
		{File: FileAttention, Sources: []string{"/spec/attention"}, Readers: []string{ServiceAttentiond}},
		{File: FileWebviewShim, Sources: []string{"/spec/attention", "/spec/apps"}, Readers: []string{ServiceWebviewShim}},
	}
}

func covers(source, path string) bool {
	return path == source || strings.HasPrefix(path, source+"/")
}

func FilesFor(path string) []string {
	files := make([]string, 0)
	for _, ownership := range FileOwnerships() {
		for _, source := range ownership.Sources {
			if covers(source, path) {
				files = append(files, ownership.File)
				break
			}
		}
	}
	return files
}

func IsOwnedPath(path string) bool {
	return len(FilesFor(path)) > 0
}

func ReadersOf(file string) []string {
	for _, ownership := range FileOwnerships() {
		if ownership.File == file {
			return ownership.Readers
		}
	}
	return nil
}

func ServicesForFiles(files []string) []string {
	seen := make(map[string]bool)
	services := make([]string, 0)
	for _, file := range files {
		for _, service := range ReadersOf(file) {
			if !seen[service] {
				seen[service] = true
				services = append(services, service)
			}
		}
	}
	sort.Strings(services)
	return services
}
